File: android/src/main/kotlin/com/loopit/minis/imgedit/ImageEditPlatformView.kt
package com.loopit.minis.imgedit

import android.content.Context
import android.view.SurfaceView
import android.view.View
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

/**
 * `loopit/minis/imgedit/canvas` factory. Each PlatformView owns its own
 * [ImageEditEngine] bound to a single Flutter viewId; the engine drives a
 * GLES 3 [SurfaceView] (the renderer in `cpp/imgedit/` handles HW-canvas
 * fallback when EGL is unavailable).
 */
class ImageEditPlatformViewFactory(
    @Suppress("unused") private val messenger: BinaryMessenger,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        @Suppress("UNCHECKED_CAST")
        val params = (args as? Map<String, Any?>) ?: emptyMap()
        val sourcePath = params["sourcePath"] as? String
        return ImageEditPlatformView(context, viewId, sourcePath)
    }
}

private class ImageEditPlatformView(
    context: Context,
    viewId: Int,
    sourcePath: String?,
) : PlatformView {

    private val surfaceView = SurfaceView(context)
    private val engine: ImageEditEngine =
        ImageEditEngine.create(viewId, surfaceView, sourcePath).also {
            ImageEditPluginRouter.register(it)
        }

    override fun getView(): View = surfaceView

    override fun dispose() {
        engine.detach()
    }
}

File: ios/Classes/ImgEdit/ImageEditPlatformView.swift
import Flutter
import Foundation
import UIKit

/// `loopit/minis/imgedit/canvas` PlatformView factory.
/// Each PlatformView creates an `ImageEditEngine` and registers it with
/// the shared `ImageEditPluginRouter` so MethodChannel calls reach it.
public final class ImageEditPlatformViewFactory: NSObject, FlutterPlatformViewFactory {
    public func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        let params = (args as? [String: Any]) ?? [:]
        let sourcePath = params["sourcePath"] as? String
        let engine = ImageEditEngine(
            viewId: viewId,
            frame: frame,
            sourcePath: sourcePath
        )
        ImageEditPluginRouter.shared.register(engine)
        return ImageEditPlatformView(engine: engine)
    }

    public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

private final class ImageEditPlatformView: NSObject, FlutterPlatformView {
    let engine: ImageEditEngine

    init(engine: ImageEditEngine) {
        self.engine = engine
        super.init()
    }

    func view() -> UIView {
        return engine.mtkView
    }
}
